### ISSUE TITLE
[pigeon] Support custom type arguments

### ISSUE THREAD

**Issue by feduke-nukem** [flutter/flutter#160503]
### Use case

Defining custom argument types would be very helpful to reduce code duplication. Nothing special about that just more Generics support

### Proposal

Opportunity to generate Apis using such classes (at least for Kotlin/Swift) 

```dart
class A<T> {
  final List<T> list;
  final T value;

  A({required this.list, required this.value});
}

sealed class Option<T> {}

class Some<T> extends Option<T> {
  final T value;

  Some({required this.value});
}

class None<T> extends Option<T> {}
```

Actually we know that in HostApi/FlutterApi only known types are used, either are generated or predefined (int, bool, String etc), so it is possible to implement such feature.

**Comment by tarrinneal** [comment]
This could end up being somewhat misleading as it wouldn't be true generic support.

It would work normally for the types that are normally supported, but not for any custom types defined outside of the pigeon file. Those types could be used, but would fail to encode/decode and crash.

I did just add sealed class extensions that could be used to solve this problem for you, but only for classes/enums.

I'll look into it more to see what I can do.

**Comment by feduke-nukem** [comment]
> It would work normally for the types that are normally supported, but not for any custom types defined outside of the pigeon file. Those types could be used, but would fail to encode/decode and crash.

Yes. Only  normally supported types and types defined within pigeon file should be allowed to be used as custom type arguments

I meant exactly that

**Comment by stuartmorgan** [comment]
Could you provide a concrete use case, so that we can evaluate the potential benefits? I'm very skeptical that we want to add ways of generating code that won't always work, and it's hard to reason about the benefits in the abstract.

**Comment by feduke-nukem** [comment]
> Could you provide a concrete use case, so that we can evaluate the potential benefits? I'm very skeptical that we want to add ways of generating code that won't always work, and it's hard to reason about the benefits in the abstract.

Let's assume that we have add-to-app application that is provided like black box.

I would like to have safely typed protocol that consumer will know about and support/implement.

For example:

```dart
@HostApi()
abstract class SomeHostApi {
  Result<SomeMethodFailure, String> getSomething();
}

sealed class Result<T, K> {}

class Success<T, K> extends Result<T, K> {
  T value;
  Success({
    required this.value,
  });
}

class Failure<T, K> extends Result<T, K> {
  K error;
  Failure({
    required this.error,
  });
}

enum SomeMethodFailure {
  noNetwork,
  unknown,
  notFound,
}
```

This is the most obvious example. In such case we got finite amount of errors that everyone is known about and should respect.

Yes we could rely on `PlatformException.code` but it is raw `String` that could contain anything.

**Comment by stuartmorgan** [comment]
I don't follow; what benefit are the generics providing in that example? It appears to be identical in practice to:

```
@HostApi()
abstract class SomeHostApi {
  SomeMethodResult getSomething();
}

sealed class SomeMethodResult {}

class SomeMethodSuccess extends SomeMethodResult {
  String value;
  StringSuccess({
    required this.value,
  });
}

class SomeMethodFailure extends SomeMethodResult {
  SomeMethodFailureCode error;
  SomeMethodFailure({
    required this.error,
  });
}

enum SomeMethodFailureCode {
  noNetwork,
  unknown,
  notFound,
}
```

**Comment by feduke-nukem** [comment]
> I don't follow; what benefit are the generics providing in that example? It appears to be identical in practice to:

In practice we should write dozens of code to follow such approach:

```dart

@HostApi()
abstract class SomeHostApi {
  Result<GetStringFailure, String> getString();

  Result<GetIntFailure, int> getInt();

  Result<GetDoubleFailure, double> getDouble();

  Result<GetBoolFailure, bool> getBool();

  Result<GetCustomFailure, Custom> getCustom();
}

class Custom {}

sealed class Result<T, K> {}

class Success<T, K> extends Result<T, K> {
  T value;
  Success({
    required this.value,
  });
}

class Failure<T, K> extends Result<T, K> {
  K error;
  Failure({
    required this.error,
  });
}

enum GetStringFailure {
  noNetwork,
  unknown,
  notFound,
}

enum GetIntFailure {
  invalidFormat,
  outOfRange,
}

enum GetDoubleFailure {
  unknownValue,
}

enum GetBoolFailure {
  somethingWentWrong,
}

enum GetCustomFailure {
  genericsAreHard,
}
```

Now image how many lines of code should we write to follow your approach

**Comment by stuartmorgan** [comment]
> In practice we should write dozens of code to follow your approach:

That code is not the code you provided before. As I said earlier:

>> Could you provide a concrete use case, so that we can evaluate the potential benefits? [...] it's hard to reason about the benefits in the abstract.

I was responding to the example you provided in response to that question.

> In the example above it would be 15 classes instead of 1 with type arguments.

The last example that you provided was apparently not an actual example of code you actually wanted to write for which you needed this feature; is this? If it's not, then the value of discussing it will be equally limited.

The goal here is not to see if it's possible to create a theoretical example where adding a (potentially problemetic) new feature would, in that theoretical case, provide value. The goal is to look at actual use cases and evaluate them, per my initial comment.

**Comment by feduke-nukem** [comment]
> > In practice we should write dozens of code to follow your approach:
> 
> That code is not the code you provided before. As I said earlier:
> 
> >> Could you provide a concrete use case, so that we can evaluate the potential benefits? [...] it's hard to reason about the benefits in the abstract.
> 
> I was responding to the example you provided in response to that question.
> 
> > In the example above it would be 15 classes instead of 1 with type arguments.
> 
> The last example that you provided was apparently not an actual example of code you actually wanted to write for which you needed this feature; is this? If it's not, then the value of discussing it will be equally limited.
> 
> The goal here is not to see if it's possible to create a theoretical example where adding a (potentially problemetic) new feature would, in that theoretical case, provide value. The goal is to look at actual use cases and evaluate them, per my initial comment.

I have provided examples are the exact case from the real world add-to-app big application video editing sdk.

It is not "theoretical" but practical. You have the "solution" to my small example so I provided a more complex scenario with the same exact problem. I actually don't quite get the request of explaining what problem generics actually solve.

We are using pigeon as a default tool to make add-to-app experience less boilerplate and more clean. So as I said, as SDK the application we would like to have more generic data structures or/and methods.

I don't know what else I should say.

**Comment by stuartmorgan** [comment]
> I actually don't quite get the request of explaining what problem generics actually solve.

For the examples you have provided, having the generator require the generics to be fully specified in all of the `@*Api` classes would be viable, which eliminates the crash-at-runtime concern; that was not clear from the initial proposal, which seemed to be for arbitrary generics.

Updating the title to reflect that narrower case, which would be reasonable to have support for.

**Comment by feduke-nukem** [comment]
> > I actually don't quite get the request of explaining what problem generics actually solve.
> 
> For the examples you have provided, having the generator require the generics to be fully specified in all of the `@*Api` classes would be viable, which eliminates the crash-at-runtime concern; that was not clear from the initial proposal, which seemed to be for arbitrary generics.
> 
> Updating the title to reflect that narrower case, which would be reasonable to have support for.

Exactly. The only supported types should be either built in such as int, string etc or types declared within the pigeon file.

That was specified above in discussion. I must apologize for not providing it in the title.

Actually we have already implemented something for that feature as a draft so I will be happy to participate.
